// cdk-dia/src/diagram/component/customizable-attribute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollapseTypes = exports.CollapssingCustomizer = void 0;
const component_1 = require("./component");
class CustomizableAttribute {
}
class CollapssingCustomizer extends CustomizableAttribute {
    constructor(type) {
        super();
        this.type = type;
    }
    static fromAttributeValue(value) {
        if (!(value in CollapseTypes))
            throw Error(value + " is not a valid enum value");
        return new CollapssingCustomizer(CollapseTypes[value]);
    }
    customize(component) {
        component.tags.set(component_1.ComponentTags.collapssingOverride, this.type);
    }
}
exports.CollapssingCustomizer = CollapssingCustomizer;
var CollapseTypes;
(function (CollapseTypes) {
    CollapseTypes["FORCE_COLLAPSE"] = "FORCE_COLLAPSE";
    CollapseTypes["FORCE_NON_COLLAPSE"] = "FORCE_NON_COLLAPSE";
    CollapseTypes["FORCE_NON_COLLAPSE_RECURSIVE"] = "FORCE_NON_COLLAPSE_RECURSIVE";
})(CollapseTypes = exports.CollapseTypes || (exports.CollapseTypes = {}));
