// cdk-dia/src/diagram/component/root-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RootComponent = void 0;
const component_1 = require("./component");
class RootComponent extends component_1.Component {
    constructor(id) {
        super(id);
        this.depth = () => 0;
        this.label = [id];
    }
    treeRoot() {
        return this;
    }
    parent() {
        throw new Error("Root component has no parent");
    }
    assureIdDoesNotExistInConnectedComponents(id) {
        const ids = this.subTreeComponentIds();
        ids.add(this.id);
        if (ids.has(id))
            throw new Error(`Id ${id} already exists in diagram`);
    }
    selfAndAllSubcomponentIds() {
        return undefined;
    }
    toSimpleObject() {
        return {
            id: this.id,
            "_image": this.icon,
            isGroup: this.subComponents().length > 0,
            children: Array.from(this.subComponents()).map(it => { return it.toSimpleObject(); })
        };
    }
    destroyLinks() {
        throw new Error("root destruction is not implemented");
    }
}
exports.RootComponent = RootComponent;
